### ISSUE TITLE
ComponentUtils.escapeJQueryId() method escapes too much for JQuery 2.4

### ISSUE THREAD

**Issue by AndrewShwaika** [primefaces/primefaces#2395]
# Extended layout with non full  page  is not working properly

<pe:layout full="false"> has been broken

ComponentUtils.escapeJQueryId() escapes every separator ':' with 4 backslashes '\\\\\\\\'
Javascript widget gets parameter with 2 backslashes. As example: #formId\\\\:clientId

JQuery 2.4 selector $('#formId\\\\:clientId') can't find an object with such id, but $('#formId\\:clientId') works

## 1) Environment
- PrimeFaces version: 6.1.1
- Application server + version: Glassfish 4.1.1, Payara 4.1.1-171
- Affected browsers: All
- PrimeFaces Extensions: 6.1.1-SNAPSHOT

## 2) Expected behavior

<pe:layout> should be visible

## 3) Actual behavior

<pe:layout> is not visible

## 4) Steps to reproduce

Any page with <pe:layout full="false">

**Comment by tandraschko** [comment]
Is there any article about those jquery change available on the web?

**Comment by AndrewShwaika** [comment]
Just debug of not working `<pe:layout>` after upgrade to Primefaces 6.1.1 and Extensions 6.1.1-SNAPSHOT.
JQuery selector could't find component with escaped string with 2 backslashes.
PrimeFaces.escapeClientId() Javascript method escapes with 1 backslash in result string and it works perfectly.

**Comment by tandraschko** [comment]
Cant replicate.
I just opnend the showcase on my local machine (which runs with jquery 2.2.4)
http://localhost:8080/showcase/ui/input/inputTextarea.xhtml

and print the following in the console:

$('#j_idt87\\\\:j_idt89').length -> 1
$('#j_idt87\\:j_idt89').length -> 0

**Comment by AndrewShwaika** [comment]
But when use ComponentUtils.escapeJQueryId() as `<pe:layout full="false">` for target.
Query will be like $('#j_idt87\\\\:j_idt89').length -> 0

**Comment by AndrewShwaika** [comment]
Issue arise when JQuery identifier formats on server side.
It comes to client side with extra backslashes in string.

**Comment by tandraschko** [comment]
Then we should check who appenss the additional backslash. IMO on the server side it looks ok.

**Comment by AndrewShwaika** [comment]
Lets count.

Replacement constant string in ComponentUtils.escapeJQueryId() has **8** backslashes.
So internal string will have **4** backslashes.
Javascript for widget creation will have **4** backslashes in constant string in config.
Internal string for configuration parameter gets **2** backslashes!

This string is directly used in JQuery selector (PrimeFaces.widget.ExtLayout Javascript fragment):

> this.jq = $(cfg.forTarget);

So jq.length -> 0

Same issue for PrimeFaces.widget.Layout

> var dialog = $(this.cfg.center.paneSelector)...

**Comment by tandraschko** [comment]
I see. You confused me with your statement that currently 2 backslashes will be rendered and doesn't work.

**Comment by tandraschko** [comment]
@melloware Do you have time to verify that PE layout is currently broken and this fixes it? :)

**Comment by melloware** [comment]
I will test this out this weekend and get back to you!

**Comment by melloware** [comment]
@tandraschko Tested and everything looks good to me.  I don't see any errors.

**Comment by tandraschko** [comment]
So it already works without this PR?

**Comment by melloware** [comment]
No i used this PR and patched my local and tested.  I thought that is what you wanted.  I can test it without the patch.  I misunderstood what you wanted me to test.

**Comment by melloware** [comment]
More details here:  https://github.com/primefaces-extensions/primefaces-extensions.github.com/issues/463

**Comment by tandraschko** [comment]
@mertsincan I applied the PR to fix this issue, the details are here:
primefaces-extensions/primefaces-extensions.github.com#463
Would you merge it to elite to fix it?

**Comment by mertsincan** [comment]
Thanks a lot for the PR ;)

**Comment by mertsincan** [comment]
Hi @tandraschko,
Reverted this PR on 6.1.5 and 6.0.23. Issue; #2653
Our PRO Users have some problems related to this PR after upgrading PrimeFaces elite version. p:autocomplete, p:wizard, p:layout etc are not working after PF6.1.3. When I  send a custom jar after reverting this PR, they say our codes work fine. Therefore, I reverted this PR on Elite versions for now. Also, one of our PRO User see the following error on console with 6.1.4;
```
pe:timePicker (Error: Syntax error, unrecognized expression: unsupported pseudo....)
```
He says this issue is fixed with your custom jar(without this PR). I wanted a sample code and maven project. When he sends it,  I will send it to you.

**Comment by tandraschko** [comment]
@melloware @Rapster we also did something in PFE, right? Could you check that?

**Comment by melloware** [comment]
We ended up not having to fix anything in PFE.  It was all fixed by this ComponentUtils fix after the Jquery Upgrade to 2.2.4.  I feel like this fix is the right fix and we need to figure out why p:wizard and p:layout are not working with this fix.

**Comment by mertsincan** [comment]
Hi @tandraschko , @melloware and @Rapster 

I attached a sample maven project. Could you please try it? An issue is reported by PRO User related to this PR.

[primefaces-test.zip](https://github.com/primefaces/primefaces/files/1241900/primefaces-test.zip)

6.2-SNAPSHOT;
- It doesn't work with this PR
- it works fine without this PR

**Comment by melloware** [comment]
I have confirmed your findings I get "Uncaught Error: Syntax error, unrecognized expression: unsupported pseudo: j_idt6"

in pe:timepicker it is doing this..

```java
writer.write(",button:'" + org.primefaces.util.ComponentUtils.escapeJQueryId(clientId)
                  + " .pe-timepicker-trigger'");
```

**Comment by melloware** [comment]
@mertsincan here is what is strange.  I am running the PFE Showcase which has a TimePicker and Layout examples using PF-6.2 SNAPSHOT exactly like your test above but everything is working fine.  Could it be that somehow in my showcase i am including more componoents on the page which is bringing more JS files in and your example somehow is missing JS files because it ONLY has a timepicker on the page?

I know there was work done in 6.2-SNAPSHOT to segment JQuery scripts even further is it possible one is not being included?

**Comment by melloware** [comment]
More research:   On the failing test this is what is printed out.

`"button:'#frmTimePicker\:timePicker .pe-timepicker-trigger'"`

But in the PFE Showcase for some reason it is not including the form which is why it is working.

`"button:'#j_idt41 .pe-timepicker-trigger'"`

I don't know why one is including the whole form and the other is not?

**Comment by AndrewShwaika** [comment]
I think extra unescape happens in rather complex fgtimepicker options setting.

**Comment by mertsincan** [comment]
@melloware, could you please try it after removing  prependId attribute of <h:form> tags on PrimeFaces Extensions Showcase?
![image](https://user-images.githubusercontent.com/11868120/29567888-3a39f010-8757-11e7-80a5-225e249d1b4e.png)

**Comment by melloware** [comment]
Ahh that's its.  PrependId was why and now I have the same issue.

**Comment by AndrewShwaika** [comment]
We have to distinguish these cases:

1. String with escaped identifier used directly in JQuery selector
1. String with escaped identifier passed to complex initialization with extra unescaping

Reverting PR will broke case 1. Identifier with 2 backslashes before ':' in string won't match any nodes.

**Comment by tandraschko** [comment]
just my 2 cents: AFAIR both versions worked fine - not sure whats wrong with layout or timepicker. But currently i don't have time for it.

**Comment by mertsincan** [comment]
@tandraschko Did you try my sample maven project?

**Comment by melloware** [comment]
So i think the real issue is with Jquery 2.2.4 that what used to be valid...

JQuery 1.X selector (2 slashes)
`$('#formId\\:clientId') `

JQuery 2.X (1 slash)
`$('#formId\:clientId')`

But because each component does things slightly differently I think we need to leave the code reverted and fix it everywhere it is wrong.

**Comment by tandraschko** [comment]
Hmm, stupid case as @AndrewShwaika already mentionend.

If the ID contains a ':', the current solution doesn't work.... If it doesn't contain the ':', it works...
The opposite happens if we revert the PR.

Same for p:layout and pe:timepicker. We must find a solution to support both cases.

**Comment by tandraschko** [comment]
Lets compare the Layout case, which uses widgetBuilder#append and results in a simple id selector $(var):
- Escaped with **4** slashes, it results on the client in #form:j_idt23, which is wrong
- Escaped with **8** slashes, it results on the client in #form\\:j_idt23, which is correct

**Comment by tandraschko** [comment]
Draggable uses widgetBuilder#attr and resuls in a complex query $(var .someClass):

- Escaped with 4 slashes, it results on the client in #form\:j_idt23, which is correct
- Escaped with 8 slashes, it results on the client in #form\\:j_idt23, which is wrong

**Comment by tandraschko** [comment]
@mertsincan 
As already stated:
JQuery 1.X selector requires **2** real slashes
$('#formId\\\\:clientId')

JQuery 2.X selector requires **1** real slash
$('#formId\\\:clientId')

So, therefore we have to reduce the escaping from **2** real slashes to **1** real slash, which happenend in this PR. This is absolutely correct for me!

The problem now happens with your commit here: https://github.com/primefaces/primefaces/commit/64fd1f2ae993e26c27d1d2e0ac77f6a97e4eeb6d
With this commit, every jqueryId rendered via wigetBuilder#attr is now escaped again and **1** real slash results in **0** real slash.

The problem with this is: some components use widgetBuilder#attr, some widgetBuilder#append, some may even use the responseWriter directly.
As i don't why you did this fix, im not sure whats the perfect fix.

**Comment by melloware** [comment]
OK so thanks to @tandraschko we have narrowed it down.  I think we need to leave the fix in for 6.2 and fix two components to work with Jquery 2.2.4/ComponentUtils...

1. PF Layout
2. PFE TimePicker

They seem to be the only two components having issues with this fix for Jquery 2.2.4.  Have I summarized that correctly?

**Comment by melloware** [comment]
OK I just tested this in PFE showcase.

1. if you use WidgetBuilder.attr such as wb.attr("forTarget", ComponentUtils.escapeJQueryId(clientId)); this fix works fine.

2. If you are using ResponseWriter directly then it does not work and you need to use the 8 slash version for it to work.

So really the only components that need fixing are ones using ResponseWriter they must escape with 8 slashes OR find a way to use WidgetBuilder.

**Comment by tandraschko** [comment]
I'm working on a fix currently. I think we should not use wb#attr and introduce something like wb#selectorAttribute

**Comment by tandraschko** [comment]
@melloware if you have gtalk or something, we can also hangup a little bit. It's sometimes much easier :D

**Comment by tandraschko** [comment]
Commited something.
The case was a little bit different. As melloware already mentionend, the ResponseWriter already escapes something and the new introcuded escapeEcmaScriptText doubles the slashes.
Therefore i reverted to the old 8 slashed and introduced a new widgetBuilder method to pass selectors, which does _NOT_ double the slashes.

**Comment by melloware** [comment]
I like your fix.  I will update PFE to use the new selectorAttribute.

**Comment by tandraschko** [comment]
Great!

**Comment by Rapster** [comment]
Hey sorry I couldn't help... I'm quite busy integrating Barcelona into our web apps. Will be quite busy for the next months :-/ Glad to hear that issue was fixed, oneof my colleague told me about this problem... Great job everybody! ;-)

**Comment by tandraschko** [comment]
Everthing is ok - you do much more as the most users! ;)

**Comment by mertsincan** [comment]
@tandraschko, I send an email about https://github.com/primefaces/primefaces/commit/64fd1f2ae993e26c27d1d2e0ac77f6a97e4eeb6d

**Comment by Rapster** [comment]
@tandraschko you mentionned gtalk but there is also https://gitter.im/ that I think will fit more our need ;-)

**Comment by tandraschko** [comment]
@Rapster We can also do that, no problem. I just used gtalk in the past as gmail is always open in the browser and i usually didn't chat with multiple persions till now.

**Comment by tandraschko** [comment]
I think the fix is also ok when e.g. PFS is used in p:block block attribute as the user must only use 1 slash:
block="@(#j_idt87\\:pnl)"

**Comment by mertsincan** [comment]
Thanks, @tandraschko. I'll add the fixes in the DraggableRenderer.java and BaseMenuRenderer.java into elite versions. Also, can we add 1 slash before ":" character in core.expressions.js for PFS?  

My test page; http://localhost:8080/showcase/ui/ajax/selector.xhtml
$('.panel :input') works
$('.panel \\:input') works
$('.ui-panel \\:input\\:not(select)') works

**Comment by tandraschko** [comment]
+1 if that works (but really only for PFS - for a normal id, document.getElementId instead of jQuery is used)

**Comment by mertsincan** [comment]
Yes, only for PFS. Also, please see "Special characters in CSS" section in the https://mathiasbynens.be/notes/css-escapes

**Comment by melloware** [comment]
@mertsincan Can you roll all of these changes into Elite 6.1.5 please?

**Comment by tandraschko** [comment]
@mertsincan Whats the important part in your link?
I think we can even do it better:
- ":" escaping should be done directly in WidgetBuilder#selectorAttr
- Components like p:block, which uses SEF#resolveClientId should also use #selectorAttr instead of #attr

WDYT?

**Comment by melloware** [comment]
@tandraschko I think you are suggesting similar to how @mertsincan added context.getResponseWriter().write(ComponentUtils.escapeEcmaScriptText(value)); in the attr() function that you would do the Jquery escaping in selectorAttr() method instead of us having to call ComponentUtils before passing into selectorAttr().

I like that.

**Comment by tandraschko** [comment]
exactly yes. This will work fine for our case now.

The second part would be if we can always escape every ":" from the SEF#resolveClientIds, which may also contain PFS.

**Comment by mertsincan** [comment]
> @mertsincan Whats the important part in your link?

About my PFS suggestion (https://mathiasbynens.be/notes/css-escapes);

>... Theoretically, the : character can be escaped as \:, but IE < 8 doesn’t recognize that escape sequence correctly...

I think the solution is good for now. We just need to be careful when we use selectorAttr method.

**Comment by mertsincan** [comment]
@melloware, I'll add this fix to elite versions.

**Comment by tandraschko** [comment]
+1
is think the current behavior with PFS is ok, users must also use a slash when they use jQuery native.

**Comment by tandraschko** [comment]
Maybe we should just think about moving the escape call inside #selectorAttr.

**Comment by melloware** [comment]
@tandraschko +1 for moving the escape inside selectorAttr.  I would think the only reason to use this special selectorAttr() call of WidgetBuilder is because you have a Jquery ID.  So why count on the caller to escape it first when that is the point of this new method.

**Comment by tandraschko** [comment]
commited something, please have a look at it :)

**Comment by melloware** [comment]
I love it.  Fixing PFE now!

**Comment by melloware** [comment]
@mertsincan I have tested all of the changes @tandraschko has made for this fix.  Can this entire fix be put into Elite 6.1.5 Release please?

**Comment by mertsincan** [comment]
Added to 6.0.23 and 6.1.5. Thanks a lot!
